File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/backend/bahnpark/model/ParkingStatus.java
package de.deutschebahn.bahnhoflive.backend.bahnpark.model;

import androidx.annotation.NonNull;
import androidx.annotation.StringRes;

import de.deutschebahn.bahnhoflive.R;
import de.deutschebahn.bahnhoflive.model.parking.ParkingFacility;
import de.deutschebahn.bahnhoflive.ui.Status;

public enum ParkingStatus {
    ALWAYS_OPEN(R.string.parking_occupancy_24_7),
    AVAILABILITY_VERY_LOW(R.string.parking_occupancy_very_low),
    AVAILABILITY_LOW(R.string.parking_occupancy_low),
    AVAILABILITY_MEDIUM(R.string.parking_occupancy_medium),
    AVAILABILITY_HIGH(R.string.parking_occupancy_high),
    CLOSED(R.string.parking_occupancy_closed, Status.NEGATIVE),
    UNKNOWN(R.string.empty, Status.UNKNOWN);

    @StringRes
    public final int label;

    public final Status status;

    private static final ParkingStatus[] VALUES = values();

    ParkingStatus(@StringRes int label) {
        this(label, Status.POSITIVE);
    }

    ParkingStatus(int label, Status status) {
        this.status = status;
        this.label = label;
    }

    public static ParkingStatus get(BahnparkSite bahnparkSite) {
        if (bahnparkSite.isParkraumIsAusserBetrieb()) {
            return CLOSED;
        }

        final BahnparkOccupancy occupancy = bahnparkSite.getOccupancy();
        if (occupancy == null) {
            return ALWAYS_OPEN;
        }

        return VALUES[Math.min(occupancy.getCategory(), VALUES.length - 2)];
    }

    @NonNull
    public static ParkingStatus get(ParkingFacility parkingFacility) {
        return UNKNOWN;
    }
}

File: modules/core/src/main/java/de/deutschebahn/bahnhoflive/backend/bahnpark/model/ParkingStatus.kt
package de.deutschebahn.bahnhoflive.backend.bahnpark.model

import androidx.annotation.StringRes
import de.deutschebahn.bahnhoflive.R
import de.deutschebahn.bahnhoflive.model.parking.ParkingFacility
import de.deutschebahn.bahnhoflive.ui.Status

enum class ParkingStatus(
    @field:StringRes val label: Int,
    val status: Status = Status.POSITIVE
) {
    ALWAYS_OPEN(R.string.parking_occupancy_24_7),
    AVAILABILITY_VERY_LOW(R.string.parking_occupancy_very_low),
    AVAILABILITY_LOW(R.string.parking_occupancy_low),
    AVAILABILITY_MEDIUM(R.string.parking_occupancy_medium),
    AVAILABILITY_HIGH(R.string.parking_occupancy_high),
    CLOSED(R.string.parking_occupancy_closed, Status.NEGATIVE),
    UNKNOWN(R.string.empty, Status.NONE);

    companion object {
        private val VALUES =
            values()

        @JvmStatic
        operator fun get(bahnparkSite: BahnparkSite): ParkingStatus {
            if (bahnparkSite.isParkraumIsAusserBetrieb) {
                return CLOSED
            }
            val occupancy = bahnparkSite.occupancy ?: return ALWAYS_OPEN
            return VALUES[Math.min(
                occupancy.category,
                VALUES.size - 2
            )]
        }

        @JvmStatic
        operator fun get(
            parkingFacility: ParkingFacility?
        ) = when {
            parkingFacility?.isOutOfService == true -> CLOSED
            parkingFacility?.liveCapacity != null -> parkingFacility.liveCapacity.parkingStatus
            else -> UNKNOWN
        }
    }

}